ScreenRecorderLib/internal_recorder.cpp

```diff
@@ -237,44 +237,67 @@ void internal_recorder::SetMaxVideoHeight(UINT32 value)
 	m_MaxVideoHeight = value;
 }
 
-std::vector<BYTE> internal_recorder::MixAudio(std::vector<BYTE> &first, std::vector<BYTE> &second)
+mix_data internal_recorder::MixAudio(std::vector<BYTE>& first, std::vector<BYTE>& second, mix_data const& prevMixData, bool const isLastCall)
 {
-	std::vector<BYTE> newvector;
+	using buffer_type = std::vector<BYTE>;
+	using size_type = buffer_type::size_type;
 
-	size_t smaller;
+	// Pre-pad the buffers with the data left over from the previous iteration
+	first.insert(first.begin(), prevMixData.firstLeftover.begin(), prevMixData.firstLeftover.end());
+	second.insert(second.begin(), prevMixData.secondLeftover.begin(), prevMixData.secondLeftover.end());
 
-	if (first.size() >= second.size())
+	// Initialize the output buffer with the data from the longer one
+	buffer_type newBuffer;
+	bool isFirstLonger = first.size() > second.size();
+	buffer_type& longerBuffer = isFirstLonger ? first : second;
+	auto const longerBufferSize = longerBuffer.size();
+
+	buffer_type& shorterBuffer = isFirstLonger ? second : first;
+	auto const shorterBufferSize = shorterBuffer.size();
+	newBuffer.resize(shorterBufferSize);
+
+	// Keep a ref to the shorter buffer
+	for (size_type i = 0; i < shorterBufferSize; i += 2) {
+		short firstSample = static_cast<short>(first[i] | first[i + 1] << 8);
+		short secondSample = static_cast<short>(second[i] | second[i + 1] << 8);
+
+		auto out = reinterpret_cast<short*>(&newBuffer[i]);
+		*out = firstSample / 2 + secondSample / 2;
+	}
```

Comment:
I'm pretty sure there was an overflow into the sign bit in the previous implementation, and this just reads a hell of a lot easier.